### PR TITLE
[dy] Write new pipeline_run_ended amplitude event

### PR DESCRIPTION
**Pull request by dy46**
# Description

Add new amplitude event `pipeline_run_ended` when a pipeline run completes in Mage. In order to add this event, I also added a new `started_at` field for pipeline runs.

The event schema in Amplitude looks like this:
Screenshot 2023-08-22 at 12 01 07 PM

# How Has This Been Tested?

- [x] Tested locally with mage-api-staging

# Checklist
- [x] The PR is tagged with proper labels (bug, enhancement, feature, documentation)
- [x] I have performed a self-review of my own code
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

cc: @wangxiaoyou1993 @mattppal

### CODE REVIEW

**Comment by mattppal** [comment]
Where is `started_at`? Was that added as a property?

**Comment by dy46** [comment]
@mattppal `started_at` is not a property on the event. it's just something I added to the data model. It's used to calculate the `run_time_seconds`.

File: mage_ai/usage_statistics/logger.py [hunk 1]
```diff
@@ -64,28 +69,76 @@ async def pipelines_impression(self, count_func: Callable) -> bool:
...
+        return await self.__send_message(
+            dict(
+                object=EventObjectType.USER,
+                action=EventActionType.IMPRESSION,
+                users=User.query.count(),
+            ),
+        )
+
+    @safe_db_query
+    async def pipeline_run_ended(self, pipeline_run: PipelineRun) -> bool:
```
Comment (wangxiaoyou1993) [inline mage_ai/usage_statistics/logger.py:94]: nit: add docstring

File: mage_ai/usage_statistics/logger.py [hunk 2]
```diff
@@ -64,28 +69,76 @@ async def pipelines_impression(self, count_func: Callable) -> bool:
...
+            pipeline_uuid=pipeline.uuid,
+            run_time_seconds=run_time_seconds,
+            trigger_method=pipeline_run.pipeline_schedule.schedule_type,
+            unique_block_types=list(set([b.get('type') for b in block_configs])),
+            unique_languages=list(set([b.get('language') for b in block_configs])),
+        )
+
+        return await self.__send_message(
+            data,
+            event_name='pipeline_run_ended',
```
Comment (wangxiaoyou1993) [inline mage_ai/usage_statistics/logger.py:129]: Can we have enum for event name?

File: mage_ai/usage_statistics/logger.py [hunk 3]
```diff
@@ -64,28 +69,76 @@ async def pipelines_impression(self, count_func: Callable) -> bool:
...
+        )
 
-    async def __send_message(self, object_name: str, action_name: str, data: Dict = None) -> bool:
+    async def __send_message(
+        self,
+        data: Dict,
+        event_name: str = 'usage_statistic.create',
```
Comment (wangxiaoyou1993) [inline mage_ai/usage_statistics/logger.py:135]: Same here. Enum for event name.

File: mage_ai/usage_statistics/logger.py [hunk 4]
```diff
@@ -64,28 +69,76 @@ async def pipelines_impression(self, count_func: Callable) -> bool:
...
+        run_time_seconds = completed_at.timestamp() - started_at.timestamp()
+
+        block_configs = pipeline.all_block_configs
+
+        data = dict(
+            num_pipeline_blocks=len(block_configs),
+            pipeline_run_uuid=pipeline_run.id,
+            pipeline_status=pipeline_run.status,
+            pipeline_type=pipeline_type,
+            pipeline_uuid=pipeline.uuid,
```
Comment (wangxiaoyou1993) [inline mage_ai/usage_statistics/logger.py:120]: Let's hash the pipeline uuid

**Comment by dy46** [comment]
@wangxiaoyou1993 addressed your comments

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**